### PR TITLE
Life cycle refactor 2

### PR DESCRIPTION
**Pull request by andorp**

### CODE REVIEW

**Comment by andorp** [comment]
I think this functionality should be discussed further.

File: src/LifeCycle.hs [hunk 1]
```diff
@@ -137,9 +150,24 @@ userCapabilities = \case
...
-
-ideaCapabilities :: AUID User -> Role -> Idea -> Phase -> [Capability]
-ideaCapabilities = phaseCap
+ideaCapabilities :: AUID User -> Role -> Idea -> Phase -> [Clickable Capability]
+ideaCapabilities u r i p =
+    let activeCaps = nub $ ideaCapabilitiesInPhase u r i p
+        allCaps    = nub $ ideaCapabilitiesInAllPhases u r i
+    in (Clickable <$> activeCaps) <> (GrayedOut <$> (allCaps \\ activeCaps))
+
+-- ASSUMPTION: The list of a phase should be part of the AllPhase list.
```
Comment (fisx) [inline src/LifeCycle.hs]: Can you rephrase this assumption?  My best guess is you mean this (but I suspect I got it wrong):

```haskell
-- The union of the dea caps of all phases.
```

File: tests/LifeCycleSpec.hs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import LifeCycle
+
+
+spec :: Spec
+spec = do
+    tag Large . it "sameItemDoesNotAppear as clickable and grayed out in ideaCapabilities" . property .
+        forAllShrinkDef arb $ \userId ->
+        forAllShrinkDef arb $ \role ->
+        forAllShrinkDef arb $ \idea ->
+        forAllShrinkDef arb $ \phase ->
```
Comment (fisx) [inline tests/LifeCycleSpec.hs]: you can probably make this one function in a 4-tuple?  if you want to, you can define `curry4` in `Types` to make this more compact.  (-:

Comment (fisx) [inline tests/LifeCycleSpec.hs]: (`uncurry4`?  never get these two right the first time..)

File: src/Frontend/Page/Idea.hs [hunk 3]
```diff
@@ -169,6 +169,13 @@ linkToIdeaLocation idea = do
              IdeaLocationSpace{} -> "Zum Ideenraum"
              IdeaLocationTopic{} -> "Zum Thema"
 
+-- ASSUMPTION: The caps list does con contain the same value marked as
+-- Clickable and GrayedOut
```
Comment (fisx) [inline src/Frontend/Page/Idea.hs]: i would prefer if this assumption is encoded as an assertion.  (can the test suite then use this assertion to fail in case of error, rather than their independent test function?)

Comment (fisx) [inline src/Frontend/Page/Idea.hs]: i tried this: a4ca3e7

Comment (fisx) [inline src/Frontend/Page/Idea.hs]: Also, I'm wondering if there is a better name?  `canGrayOut`?  No, that's no better.  I appreciate the rule that eliminators are called like the type, `maybe` does it, too. I just still find it hard to read.

**Comment by fisx** [comment]
Let's discuss my comments and changes off-line, but I think this can be merged soon.

File: src/LifeCycle.hs (hunk 1 above)
Comment (andorp) [inline src/LifeCycle.hs]: thanks

File: tests/LifeCycleSpec.hs (hunk 2 above)
Comment (andorp) [inline tests/LifeCycleSpec.hs]: I think in this form we run 4x more tests, but I am not insisting on that.

Comment (fisx) [inline tests/LifeCycleSpec.hs]: not important, leave it.

File: src/Frontend/Page/Idea.hs (hunk 3 above)
Comment (fisx) [inline src/Frontend/Page/Idea.hs]: i like, in this order:

1. eitherClickable
1. eitherClickableGrayedOut
1. clickable

all are ok for me.

**Comment by fisx** [comment]
@andorp i've played with your changes, add a few more, and opened #579.  could you take a last look at my commits here and object or merge?  thanks!

**Comment by fisx** [comment]
Ok, changed my mind: this is making the code base too complicated for too little benefit.  If we want to have a bad solution for some grayed-out UI elements here and there, we should aim for a *simple*, *local* bad solution, not something striving for generality and beauty unsuccessfully.

I will close the issue for now, but leave the PR open in the hope we'll find the time to think of something better.

**Comment by andorp** [comment]
Closing this, as we closed other experiments in the past.
